File: lib/integrations/adapters/vercel.ts
/**
 * Thin Vercel REST API adapter.
 *
 * Reads (listProjects, getEnvVars) execute directly. Writes (setEnvVar,
 * triggerDeployment) are MUTATING — callers must gate them through the
 * approval flow before invoking. This file does not enforce that gate;
 * the runtime that wires the tool does.
 *
 * Auth: IntegrationConnection (toolkit='vercel', status='active') →
 * VERCEL_TOKEN env. Throws when neither is set.
 *
 * Env values are NEVER returned from getEnvVars; we report set/unset only.
 */

import { supabase } from '@/lib/supabase';

const VERCEL_API = 'https://api.vercel.com';

async function getVercelToken(spaceId: string): Promise<string> {
  const { data } = await supabase
    .from('IntegrationConnection')
    .select('accessToken')
    .eq('spaceId', spaceId)
    .eq('toolkit', 'vercel')
    .eq('status', 'active')
    .maybeSingle();

  const token = (data as { accessToken?: string } | null)?.accessToken;
  if (token) return token;

  const env = process.env.VERCEL_TOKEN;
  if (env) return env;

  throw new Error(
    `Vercel: no token found for space ${spaceId}. ` +
      `Connect Vercel in Settings → Integrations, or set VERCEL_TOKEN.`,
  );
}

function vercelHeaders(token: string): HeadersInit {
  return { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' };
}

async function readOrThrow(res: Response, action: string): Promise<unknown> {
  if (!res.ok) {
    let body: { error?: { message?: string } } = {};
    try {
      body = (await res.json()) as { error?: { message?: string } };
    } catch {
      // ignore
    }
    const msg = body.error?.message ?? (await res.text().catch(() => ''));
    throw new Error(`Vercel ${action} failed: ${res.status} — ${msg}`);
  }
  return res.json();
}

// ── Types ────────────────────────────────────────────────────────────────────

export interface VercelProject {
  id: string;
  name: string;
  framework: string | null;
}

export interface VercelEnvVar {
  key: string;
  marker: 'set' | 'unset';
  target: string[];
}

export interface VercelDeployment {
  id: string;
  url: string;
  state: string | null;
}

// ── Operations ───────────────────────────────────────────────────────────────

export async function vercelListProjects(spaceId: string): Promise<VercelProject[]> {
  const token = await getVercelToken(spaceId);
  const res = await fetch(`${VERCEL_API}/v9/projects?limit=50`, {
    headers: vercelHeaders(token),
  });
  const data = (await readOrThrow(res, 'listProjects')) as {
    projects?: Array<{ id: string; name: string; framework: string | null }>;
  };
  return (data.projects ?? []).map((p) => ({
    id: p.id,
    name: p.name,
    framework: p.framework ?? null,
  }));
}

export async function vercelGetEnvVars(
  spaceId: string,
  projectId: string,
): Promise<VercelEnvVar[]> {
  const token = await getVercelToken(spaceId);
  const res = await fetch(`${VERCEL_API}/v9/projects/${projectId}/env`, {
    headers: vercelHeaders(token),
  });
  const data = (await readOrThrow(res, 'getEnvVars')) as {
    envs?: Array<{ key: string; value?: string; type?: string; target?: string[] }>;
  };
  return (data.envs ?? []).map((e) => ({
    key: e.key,
    marker:
      e.value || (e.type && ['encrypted', 'secret', 'system'].includes(e.type)) ? 'set' : 'unset',
    target: e.target ?? [],
  }));
}

export async function vercelSetEnvVar(
  spaceId: string,
  params: { projectId: string; key: string; value: string; target?: string },
): Promise<{ id: string }> {
  const token = await getVercelToken(spaceId);
  const target = [params.target ?? 'production'];
  const res = await fetch(`${VERCEL_API}/v10/projects/${params.projectId}/env`, {
    method: 'POST',
    headers: vercelHeaders(token),
    body: JSON.stringify({
      key: params.key,
      value: params.value,
      target,
      type: 'encrypted',
    }),
  });
  const data = (await readOrThrow(res, 'setEnvVar')) as { id?: string; created?: { id?: string } };
  const id = data.id ?? data.created?.id ?? '';
  return { id };
}

export async function vercelTriggerDeployment(
  spaceId: string,
  params: { projectId: string; ref?: string },
): Promise<VercelDeployment> {
  const token = await getVercelToken(spaceId);
  // Vercel's create-deployment endpoint requires a project name; the v6
  // endpoint accepts a project id via `project` param + `gitSource`.
  const res = await fetch(`${VERCEL_API}/v13/deployments`, {
    method: 'POST',
    headers: vercelHeaders(token),
    body: JSON.stringify({
      project: params.projectId,
      gitSource: { type: 'github', ref: params.ref ?? 'main' },
    }),
  });
  const data = (await readOrThrow(res, 'triggerDeployment')) as {
    id?: string;
    url?: string;
    readyState?: string;
  };
  return {
    id: data.id ?? '',
    url: data.url ?? '',
    state: data.readyState ?? null,
  };
}

File: agent/tools/engineering/vercel.py
"""Vercel tools — list/inspect projects, gate writes through approval.

Reads (list_projects, get_env_vars) execute directly. Writes
(set_env_var, trigger_deployment) return an "ACTION REQUIRES APPROVAL"
string instead of executing — the manager turns those into approval
requests the founder confirms before re-invocation.

Auth: IntegrationConnection (toolkit='vercel') → VERCEL_TOKEN env.
"""

from __future__ import annotations

import os
from typing import Any

import httpx
import structlog

from agents import function_tool
from db import supabase

logger = structlog.get_logger(__name__)

_VERCEL_API = "https://api.vercel.com"
_TIMEOUT = 20.0


async def _get_vercel_token(space_id: str) -> str | None:
    if space_id:
        try:
            db = await supabase()
            res = await (
                db.table("IntegrationConnection")
                .select("accessToken")
                .eq("spaceId", space_id)
                .eq("toolkit", "vercel")
                .eq("status", "active")
                .maybe_single()
                .execute()
            )
            if res.data and res.data.get("accessToken"):
                return res.data["accessToken"]
        except Exception as err:  # noqa: BLE001
            logger.warning("vercel_token_lookup_failed", space_id=space_id, error=str(err)[:200])

    return os.environ.get("VERCEL_TOKEN")


def _auth_headers(token: str) -> dict[str, str]:
    return {"Authorization": f"Bearer {token}", "Content-Type": "application/json"}


def _handle_response(resp: httpx.Response, action: str) -> str | None:
    if resp.status_code == 401:
        return f"Vercel auth failed for {action}. Check token or reconnect Vercel."
    if resp.status_code == 403:
        return f"Vercel permission denied for {action}. Token scope may be insufficient."
    if not resp.is_success:
        body: dict[str, Any] = {}
        try:
            body = resp.json()
        except Exception:  # noqa: BLE001
            pass
        msg = (body.get("error") or {}).get("message") or resp.text[:200]
        return f"Vercel API error {resp.status_code} for {action}: {msg}"
    return None


@function_tool
async def vercel_list_projects(space_id: str = "") -> str:
    """List the Vercel projects the connected token can see."""
    token = await _get_vercel_token(space_id)
    if not token:
        return "No Vercel token found. Connect Vercel in settings or set VERCEL_TOKEN."

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        resp = await client.get(
            f"{_VERCEL_API}/v9/projects",
            headers=_auth_headers(token),
            params={"limit": 50},
        )

    err = _handle_response(resp, "list_projects")
    if err:
        return err

    projects = resp.json().get("projects", [])
    if not projects:
        return "No Vercel projects."

    return "\n".join(f"{p.get('id')}  {p.get('name')}  framework={p.get('framework') or 'n/a'}" for p in projects)


@function_tool
async def vercel_get_env_vars(project_id: str, space_id: str = "") -> str:
    """List env var KEYS for a Vercel project. Values are NEVER returned — only set/unset."""
    token = await _get_vercel_token(space_id)
    if not token:
        return "No Vercel token found. Connect Vercel in settings or set VERCEL_TOKEN."

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        resp = await client.get(
            f"{_VERCEL_API}/v9/projects/{project_id}/env",
            headers=_auth_headers(token),
        )

    err = _handle_response(resp, "get_env_vars")
    if err:
        return err

    envs = resp.json().get("envs", [])
    if not envs:
        return f"No env vars on project {project_id}."

    lines = []
    for e in envs:
        key = e.get("key")
        # Vercel returns 'value' only for plaintext non-secret vars; treat
        # anything we can't read as 'set' to avoid leaking secrets through
        # the agent's transcript.
        marker = "set" if e.get("value") or e.get("type") in {"encrypted", "secret", "system"} else "unset"
        targets = ",".join(e.get("target") or [])
        lines.append(f"{key} = ({marker}) target={targets}")
    return "\n".join(lines)


@function_tool
async def vercel_set_env_var(
    project_id: str,
    key: str,
    value: str,
    target: str = "production",
    space_id: str = "",
) -> str:
    """Set an env var on a Vercel project.

    ACTION REQUIRES APPROVAL — surfaces a gate string; the manager runs it
    only after founder confirmation.
    """
    return (
        f"ACTION REQUIRES APPROVAL: set env var '{key}' on Vercel project "
        f"{project_id} (target={target}). Value will be stored encrypted; "
        f"the founder must approve before this is applied."
    )


@function_tool
async def vercel_trigger_deployment(
    project_id: str,
    ref: str = "main",
    space_id: str = "",
) -> str:
    """Trigger a deployment of `ref` on the given Vercel project.

    ACTION REQUIRES APPROVAL — surfaces a gate string; the manager runs it
    only after founder confirmation.
    """
    return (
        f"ACTION REQUIRES APPROVAL: trigger deployment of {ref} on Vercel "
        f"project {project_id}. The founder must approve before this is applied."
    )
